feat(utxo-core): PayGo Attestation util functions

davidkaplanbitgo opened the pull request:
TICKET: BTC-2047

<!--
# Please be aware of the following when making your pull request:

## Description

Please include a summary of your proposed changes and which issue is being addressed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Issue Number

Internal Users - Please include the related internal tracking number (e.g. BG-000000).
External Users - Please link to any relevant github issues as necessary.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My code compiles correctly for both Node and Browser environments
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My commits follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and I have properly described any BREAKING CHANGES
- [ ] The ticket or github issue was included in the commit message as a reference
- [ ] I have made corresponding changes to the documentation and on any new/updated functions and/or methods - [jsdoc](https://jsdoc.app/)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
-->

christopher-fong committed:
chore: wip paygo util functions

TICKET: BTC-2047

chore: wip adding tests for paygo address proof util functions

TICKET: BTC-2047

chore: added testing and functions to get proprietary key vals from output

TICKET: BTC-2047

chore: wip added functions and tests

TICKET: BTC-2047

davidkaplanbitgo commented on modules/utxo-lib/src/bitgo/psbt/paygoAddressProof.ts:
```diff
@@ -0,0 +1,87 @@
+ * @param inputIndex - the index of the address in our output
+ * @param sig - the signature that we want to encode
+ */
+export function addPaygoAddressProof(psbt: UtxoPsbt, inputIndex: number, sig: Buffer): void {
+  psbt.addProprietaryKeyValToInput(inputIndex, {
```
```suggestion
export function addPaygoAddressProof(psbt: UtxoPsbt, outputIndex: number, sig: Buffer): void {
  psbt.addProprietaryKeyValToOutput(outputIndex, {
```

davidkaplanbitgo commented on modules/utxo-lib/test/bitgo/psbt/paygoAddressProof.ts:
```diff
@@ -0,0 +1,101 @@
+    });
+
+    it('should add Paygo Adress Proof to non-empty PSBT', () => {
+        const inputIndex = 0;
```
```suggestion
        const outputIndex = 0;
```

davidkaplanbitgo commented on modules/utxo-lib/test/bitgo/psbt/paygoAddressProof.ts:
```diff
@@ -0,0 +1,101 @@
+        const inputIndex = 0;
+        const psbt = getTestPsbt(psbtInputs, psbtOutputs, 'fullsigned');
+        addPaygoAddressProof(psbt, inputIndex, Buffer.from(sig));
+        const proprietaryKeyVals = psbt.data.globalMap.unknownKeyVals;
+        assert(proprietaryKeyVals);
+        const proofInPsbt = getPaygoProprietaryKey(proprietaryKeyVals);
+        assert(proofInPsbt.length === 1);
```
just do a manual extraction of the proprietary key val beign added here and dont extract that utility into a function

davidkaplanbitgo commented on modules/utxo-lib/test/bitgo/psbt/paygoAddressProof.ts:
```diff
@@ -0,0 +1,101 @@
+        assert(proofInPsbt.length === 1);
+    });
+
+    it('should add Paygo Adress Proof to non-empty PSBT', () => {
```
```suggestion
    it('should add and verify and valid paygo address proof on a PSBT', () => {
```

davidkaplanbitgo commented on modules/utxo-lib/test/bitgo/psbt/paygoAddressProof.ts:
```diff
@@ -0,0 +1,101 @@
+            return keyValue.key.identifier === PSBT_PROPRIETARY_IDENTIFIER && keyValue.value === Buffer.from(sig)
+        });
+    }
+    it('should add PayGo Address proof to empty PSBT', () => {
```
```suggestion
    it('should fail a proof verification if the proof isnt valid', () => {
```

davidkaplanbitgo commented on modules/utxo-lib/src/bitgo/psbt/paygoAddressProof.ts:
```diff
@@ -0,0 +1,87 @@
+export const PayGoAddressProofKey: ProprietaryKey = {
+  identifier: PSBT_PROPRIETARY_IDENTIFIER,
+  subtype: ProprietaryKeySubtype.PAYGO_ADDRESS_PROOF,
+  // Is there a better way to not provide a pubkey?
+  keydata: Buffer.from([]),
```
```suggestion
  // Is there a better way to not provide a pubkey?
  keydata: Buffer.from([]),
```
maybe what is encoded is the pubkey used for the verification

davidkaplanbitgo commented on modules/utxo-lib/src/bitgo/psbt/paygoAddressProof.ts:
```diff
@@ -0,0 +1,87 @@
+ */
+export function addPaygoAddressProof(psbt: UtxoPsbt, inputIndex: number, sig: Buffer): void {
+  psbt.addProprietaryKeyValToInput(inputIndex, {
+    key: PayGoAddressProofKey,
```
```suggestion
    key: {
      identifier: PSBT_PROPRIETARY_IDENTIFIER,
      subtype: ProprietaryKeySubtype.PAYGO_ADDRESS_PROOF,
      keydata: Buffer.from(pub)
  },
```

davidkaplanbitgo commented on modules/utxo-lib/test/bitgo/psbt/paygoAddressProof.ts:
```diff
@@ -0,0 +1,101 @@
+    });
+
+    // do we have an error check with the original addProprietaryKeyValToInput
+    // to see if there are duplicates of the same? if not I will add the test case here.
```
make a test that has a proof on there already, and then you add a new one and make sure it gets substituted

davidkaplanbitgo commented on modules/utxo-lib/test/bitgo/psbt/paygoAddressProof.ts:
```diff
@@ -0,0 +1,101 @@
+})
+
+
+describe('verifyPaygoAddressProof', () => {
+    const pubkey = 'pubkeytestforpaygopsbt'
+
+    it('should verify a valid PayGo address proof', () => {
+        const indexInput = 0;
+        const psbt = addPaygoAddressProofToPsbt(indexInput, false, false);
+        verifyPaygoAddressProof(psbt, 0, Buffer.from(pubkey))
+        // nothing happens if verification is successful
+    });
+
+    it('should throw an error if there is no PayGo address in PSBT', () => {
+        const psbt = getTestPsbt(psbtInputs, psbtOutputs, 'fullsigned');
+        assert.throws(() => verifyPaygoAddressProof(psbt, 0, Buffer.from(pubkey)), (e: any) => e.message === 'here is no paygo address proof encoded in the PSBT.')
+    });
+
+    it('should throw an error if there is multiple PayGo address in PSBT', () => {
+        const psbt = getTestPsbt(psbtInputs, psbtOutputs, 'fullsigned');
+        const sig2 = 'paygoaddresssig2'
+        addPaygoAddressProof(psbt, 0, Buffer.from(sig))
+        addPaygoAddressProof(psbt, 1, Buffer.from(sig2))
+        assert.throws(() => verifyPaygoAddressProof(psbt, 0, Buffer.from(pubkey)), (e: any) => e.message === 'There are multiple paygo address proofs encoded in the PSBT. Something went wrong.')
+    });
+
+    // Once we think of what message should be signing, we can generate a test case
+    // to test this error message properly
+    it('should throw an error if the verification fails', () => {
+        const psbt = addPaygoAddressProofToPsbt(0, false, false);
+        assert.throws(() => verifyPaygoAddressProof(psbt, 0, Buffer.from(pubkey + 's')), (e: any) => e.message === 'Cannot verify the paygo address signature with the provided pubkey.')  
+    });
+});
```
Not necessary if the above test suite is changed above
```suggestion
```

davidkaplanbitgo commented on modules/utxo-lib/test/bitgo/psbt/paygoAddressProof.ts:
```diff
@@ -0,0 +1,101 @@
+});
+
+describe('getPaygoAddressProofIndex', () => {
+    it('should get PayGo address proof index from PSBT', () => {
```
```suggestion
    it('should get PayGo address proof index from PSBT if there is one', () => {
```

davidkaplanbitgo commented on modules/utxo-lib/test/bitgo/psbt/paygoAddressProof.ts:
```diff
@@ -0,0 +1,101 @@
+
+describe('getPaygoAddressProofIndex', () => {
+    it('should get PayGo address proof index from PSBT', () => {
+    });
```
add test case to make sure that you return undefined if there is no address proof
add test case to fail this function if there are more than 1

davidkaplanbitgo reviewed (commented):
Good start!

christopher-fong commented on modules/utxo-lib/src/bitgo/UtxoPsbt.ts:
```diff
@@ -1142,6 +1143,73 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint> = UtxoTransaction<bigin
+   * Adds a proprietary key value pair to PSBT output
+   * Default identifier is utf-8 for identifier
+   */
+  addProprietaryKeyValToOutput(outputIndex: number, keyValueData: ProprietaryKeyValue): this {
```
I added these new functions on the UtxoPsbt class because the original proprietary key functions only retrieved them from the input map, where we wanted to get the proprietary key vals and add them in the output map

davidkaplanbitgo reviewed (commented):
Looking good!

christopher-fong committed:
chore: added test for paygo addr and moved generate paygo proof to other pr

TICKET: BTC-2047

chore: determined msg to sign and verify

Issue: BTC-2047

TICKET: BTC-2047

chore(utxo-lib): modified verify function and tests

Issue: BTC-2047

TICKET: BTC-2047

chore(utxo-core): moved util functions to uxto-core

TICKET: BTC-2047

chore: got tests to work

BTC-2047

TICKET: BTC-2047

feat(utxo-core): cleaned up comments

TICKET: BTC-2047

chore(utxo-core): added bitcoinjs-message to deps and fixed import

TICKET: BTC-2047

chore(utxo-core): used new helper functions for prop kv func and removed old

TICKET: BTC-2047

christopher-fong committed:
chore(utxo-core): add error classes for util func

TICKET: BTC-2047

chore(utxo-core): renamed functions consistent and error prototypes

TICKET: BTC-2047

chore(utxo-core): bip32utils to sign and verify message

TICKET: BTC-2047

chore(utxo-core): added bitcoinjs-message back to package.json

TICKET: BTC-2047

chore(utxo-core): refactored functions and debug tests

TICKET: BTC-2047

chore(utxo-core): update verify function

TICKET: BTC-2047

chore(utxo-core): updated test and verify

TICKET: BTC-2047

chore(utxo-core): clean up rebase conflicts

TICKET: BTC-2047

chore(utxo-core): added helper function to create attestation

TICKET: BTC-2047

OttoAllmendinger commented on modules/utxo-core/src/paygo/psbt/PayGoUtils.ts:
```diff
@@ -0,0 +1,129 @@
+ * @param message - The message we want to verify corresponding to sig
+ * @returns
+ */
+export function verifyPaygoAddressProof(
```
use `PayGo` consistently

OttoAllmendinger commented on modules/utxo-core/src/paygo/psbt/PayGoUtils.ts:
```diff
@@ -0,0 +1,123 @@
+  ErrorPayGoAddressProofFailedVerification,
+} from './Errors';
+
+const NILLUUID = '00000000-0000-0000-0000-000000000000';
```
```suggestion
const NIL_UUID = '00000000-0000-0000-0000-000000000000';
```

OttoAllmendinger commented on modules/utxo-core/test/paygo/psbt/PayGoUtils.ts:
```diff
@@ -0,0 +1,154 @@
+const attestationPrvKey = dummyPub1.user.privateKey!;
+
+// UUID structure
+const nillUUID = '00000000-0000-0000-0000-000000000000';
```
export the one from `src/` and use it here

OttoAllmendinger commented on modules/utxo-core/src/paygo/psbt/PayGoUtils.ts:
rename this to `payGoAddressProof.ts`

christopher-fong committed:
chore(utxo-core): renamed files and moved helper to own file

TICKET: BTC-2047

chore(utxo-core): updated index.ts

TICKET: BTC-2047

chore(utxo-core): added asserts for address in attestation

TICKET: BTC-2047

chore(utxo-core): naming convention

TICKET: BTC-2047

OttoAllmendinger commented on modules/utxo-core/src/paygo/attestation.ts:
```diff
@@ -0,0 +1,13 @@
+ * @param entropy
+ * @returns
+ */
+export function createPayGoAttestationBuffer(address: string, entropy: Buffer): Buffer {
+  const addressBuffer = Buffer.from(address);
+  return Buffer.concat([entropy, addressBuffer, Buffer.from(NILL_UUID)]);
+}
```
isn't entropy fixed-sized?

let's assert `address` is decodable as well (pass the `network` argument)

OttoAllmendinger commented on modules/utxo-core/test/paygo/psbt/payGoAddressProof.ts:
```diff
@@ -0,0 +1,151 @@
+  });
+});
+
+describe('createPayGoAttestationBuffer', () => {
+  it('should create a PayGo Attestation proof matching with original proof', () => {
+    const payGoAttestationProof = createPayGoAttestationBuffer(addressToVerify, addressProofEntropy);
+    assert.strictEqual(payGoAttestationProof.toString(), addressProofMsgBuffer.toString());
+    assert(Buffer.compare(payGoAttestationProof, addressProofMsgBuffer) === 0);
+  });
+
+  it('should create a PayGo Attestation proof that does not match with different uuid', () => {
+    const addressProofBufferDiffUuid = generatePayGoAttestationProof(
+      '00000000-0000-0000-0000-000000000001',
+      Buffer.from(addressToVerify)
+    );
+    const payGoAttestationProof = createPayGoAttestationBuffer(addressToVerify, addressProofEntropy);
+    assert.notStrictEqual(payGoAttestationProof.toString(), addressProofBufferDiffUuid.toString());
+    assert(Buffer.compare(payGoAttestationProof, addressProofBufferDiffUuid) !== 0);
+  });
+});
```
move this to a separate file `attestation.ts`

OttoAllmendinger commented on modules/utxo-core/test/paygo/psbt/payGoAddressProof.ts:
```diff
@@ -0,0 +1,151 @@
+    addPayGoAddressProof(psbt, outputIndex, sig, addressProofEntropy);
+    addPayGoAddressProof(psbt, outputIndex, Buffer.from('signature2'), crypto.randomBytes(64));
+    const output = checkForOutput(psbt.data.outputs, outputIndex);
+    const proofInPsbt = getPaygoProprietaryKey(output.unknownKeyVals!);
```
PayGo

OttoAllmendinger commented on modules/utxo-core/src/paygo/attestation.ts:
```diff
@@ -0,0 +1,13 @@
+export const NILL_UUID = '00000000-0000-0000-0000-000000000000';
```
```suggestion
export const NIL_UUID = '00000000-0000-0000-0000-000000000000';
```

OttoAllmendinger commented on modules/utxo-core/src/testutil/parseVaspProof.ts:
```diff
@@ -0,0 +1,25 @@
+ * @param proof
+ * @returns
+ */
+export function parseVaspProof(proof: Buffer): Buffer {
```
```suggestion
export function trimMessagePrefix(proof: Buffer): Buffer {
```

OttoAllmendinger replied:
please edit the JSDoc to be more concise

OttoAllmendinger commented on modules/utxo-core/src/paygo/psbt/Errors.ts:
```diff
@@ -0,0 +1,29 @@
+export class ErrorNoPayGoProof extends Error {
+  constructor(public outputIndex: number) {
+    super(`There is no paygo address proof encoded in the PSBT at output ${outputIndex}.`);
+  }
+}
+
+export class ErrorMultiplePayGoProof extends Error {
+  constructor() {
+    super('There are multiple paygo address proofs encoded in the PSBT. Something went wrong.');
+  }
+}
+
+export class ErrorPayGoAddressProofFailedVerification extends Error {
+  constructor() {
+    super('Cannot verify the paygo address signature with the provided pubkey.');
+  }
+}
+
+export class ErrorOutputIndexOutOfBounds extends Error {
+  constructor(public outputIndex: number) {
+    super(`Output index ${outputIndex} is out of bounds for PSBT outputs.`);
+  }
+}
+
+export class ErrorMultiplePayGoProofAtPsbtIndex extends Error {
+  constructor(public outputIndex: number) {
+    super(`There are multiple PayGo addresses in the PSBT output ${outputIndex}.`);
+  }
+}
```
if you want this to be extra nice, have them all inherit from a toplevel

```
class ErrorPayGoProof extends Error {}
```

christopher-fong committed:
chore(utxo-core): fixed pr on error class and JSDocs

TICKET: BTC-2047
